In [2]:
import pandas as pd


#df_business = pd.read_pickle(r"datasets\datasets yelps\business.pkl")
#df_checkin = pd.read_json(r"datasets\datasets yelps\checkin.json", lines= True)
df_tip = pd.read_json(r"datasets\datasets yelps\tip.json", lines= True)
#tuve que instalar pyarrow para levantar el parquet: pip install pyarrow
#df_user = pd.read_parquet(r"datasets\datasets yelps\user.parquet")

dfs = []

#cargo dataset review por chunks:

#for chunk in pd.read_json(r"datasets\datasets yelps\review.json", lines= True, chunksize=10000):

#    dfs.append(chunk)


#df_review_dividido = pd.concat(dfs, ignore_index=True)

FileNotFoundError: File datasets\datasets yelps\tip.json does not exist

# EDA básico: business dataset

In [40]:
df_business.head()

business_id                      name   
0  Pns2l4eNsfO8kk83dixA6A  Abby Rappoport, LAC, CMQ  \
1  mpf3x-BjTdTEA3yCZrAYPw             The UPS Store   
2  tUFrWirKiKi_TAnsVWINQQ                    Target   
3  MTSW4McQd7CbVtyjqoe9mw        St Honore Pastries   
4  mWMc6_wTdE0EUBKIGXDVfA  Perkiomen Valley Brewery   

                           address           city state postal_code   
0           1616 Chapala St, Ste 2  Santa Barbara   NaN       93101  \
1  87 Grasso Plaza Shopping Center         Affton   NaN       63123   
2             5255 E Broadway Blvd         Tucson   NaN       85711   
3                      935 Race St   Philadelphia    CA       19107   
4                    101 Walnut St     Green Lane    MO       18054   

    latitude   longitude stars review_count  ... state postal_code latitude   
0  34.426679 -119.711197   5.0            7  ...   NaN         NaN      NaN  \
1  38.551126  -90.335695   3.0           15  ...   NaN         NaN      NaN   
2  32.223236 -110.880452   3.5           22  ...   NaN         NaN      NaN   
3  39.955505  -75.155564   4.0           80  ...   NaN         NaN      NaN   
4  40.338183  -75.471659   4.5           13  ...   NaN         NaN      NaN   

  longitude stars review_count is_open attributes categories hours  
0       NaN   NaN          NaN     NaN        NaN        NaN   NaN  
1       NaN   NaN          NaN     NaN        NaN        NaN   NaN  
2       NaN   NaN          NaN     NaN        NaN        NaN   NaN  
3       NaN   NaN          NaN     NaN        NaN        NaN   NaN  
4       NaN   NaN          NaN     NaN        NaN        NaN   NaN  

[5 rows x 28 columns]

Veo que hay bastantes datos faltantes al menos en las primeras 5 rows además de que hay columnas repetidas

In [41]:
df_business.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 28 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
 14  business_id   5 non-null       object
 15  name          5 non-null       object
 16  address       5 non-null       object
 17  city          5 non-null       object
 18  state         5 non-null     

Vistazo general del dataset

In [42]:
df_business.dtypes

business_id     object
name            object
address         object
city            object
state           object
postal_code     object
latitude        object
longitude       object
stars           object
review_count    object
is_open         object
attributes      object
categories      object
hours           object
business_id     object
name            object
address         object
city            object
state           object
postal_code     object
latitude        object
longitude       object
stars           object
review_count    object
is_open         object
attributes      object
categories      object
hours           object
dtype: object

El tipo de dato para todas las columnas es strings teniendo datos de distintos tipos en la misma columna.
Observo que el dataset tiene todas las columnas repetidas una vez.

In [43]:
df_business.isnull().sum()

business_id          0
name                 0
address              0
city                 0
state                3
postal_code          0
latitude             0
longitude            0
stars                0
review_count         0
is_open              0
attributes       13744
categories         103
hours            23223
business_id     150341
name            150341
address         150341
city            150341
state           150341
postal_code     150341
latitude        150341
longitude       150341
stars           150341
review_count    150341
is_open         150341
attributes      150341
categories      150341
hours           150341
dtype: int64

Casi el 10% de los datos en la columna attributes son nulos.
En cuanto a la columna categories, tiene muy pocos datos faltantes.
La columna hours tiene un 15% de datos faltantes.
Finalmente, las columnas repetidas tienen casi el 100% de sus valores como nulos.

In [44]:
df_business["categories"].value_counts

<bound method DataFrame.value_counts of                                                categories categories
0       Doctors, Traditional Chinese Medicine, Naturop...        NaN
1       Shipping Centers, Local Services, Notaries, Ma...        NaN
2       Department Stores, Shopping, Fashion, Home & G...        NaN
3       Restaurants, Food, Bubble Tea, Coffee & Tea, B...        NaN
4                               Brewpubs, Breweries, Food        NaN
...                                                   ...        ...
150341                         Nail Salons, Beauty & Spas        NaN
150342  Pets, Nurseries & Gardening, Pet Stores, Hobby...        NaN
150343  Shopping, Jewelry, Piercing, Toy Stores, Beaut...        NaN
150344  Fitness/Exercise Equipment, Eyewear & Optician...        NaN
150345  Beauty & Spas, Permanent Makeup, Piercing, Tattoo        NaN

[150346 rows x 2 columns]>

En categorias hay tags que representan el tipo de local o establecimiento. La gran mayoría comprende valores únicos por lo que hay muchos tipos de establecimientos que van desde restaurantes hasta salones de belleza.

In [45]:
df_business["attributes"].value_counts

<bound method DataFrame.value_counts of                                                attributes attributes
0                           {'ByAppointmentOnly': 'True'}        NaN
1                  {'BusinessAcceptsCreditCards': 'True'}        NaN
2       {'BikeParking': 'True', 'BusinessAcceptsCredit...        NaN
3       {'RestaurantsDelivery': 'False', 'OutdoorSeati...        NaN
4       {'BusinessAcceptsCreditCards': 'True', 'Wheelc...        NaN
...                                                   ...        ...
150341  {'ByAppointmentOnly': 'False', 'RestaurantsPri...        NaN
150342  {'BusinessAcceptsCreditCards': 'True', 'Restau...        NaN
150343  {'RestaurantsPriceRange2': '1', 'BusinessAccep...        NaN
150344  {'BusinessParking': '{'garage': False, 'street...        NaN
150345  {'WheelchairAccessible': 'True', 'BusinessAcce...        NaN

[150346 rows x 2 columns]>

La columna attributes contiene informacion sobre caracteristicas del establecimiento.

# EDA básico: review dataset (fragmento del dataset original)

In [46]:
df_review_dividido.head()

review_id                 user_id             business_id   
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw  \
1  BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q  7ATYjTIgM3jUlt4UM3IypQ   
2  saUsX_uimxRlCVr67Z4Jig  8g_iMtfSiwikVnbP2etR0A  YjUWPpI6HXG530lwP-fb2A   
3  AqPFMleE6RsU23_auESxiA  _7bHUi9Uuf5__HHc_Q8guQ  kxX2SOes4o-D3ZQBkiMRfA   
4  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   

   stars  useful  funny  cool   
0      3       0      0     0  \
1      5       1      0     1   
2      3       0      0     0   
3      5       1      0     1   
4      4       1      0     1   

                                                text                date  
0  If you decide to eat here, just be aware it is... 2018-07-07 22:09:11  
1  I've taken a lot of spin classes over the year... 2012-01-03 15:28:18  
2  Family diner. Had the buffet. Eclectic assortm... 2014-02-05 20:30:30  
3  Wow!  Yummy, different,  delicious.   Our favo... 2015-01-04 00:01:03  
4  Cute interior and owner (?) gave us tour of up... 2017-01-14 20:54:15

In [47]:
df_review_dividido.shape

(6990280, 9)

Casi 7 millones de rows y 9 columnas

In [48]:
df_review_dividido.describe()

stars        useful         funny          cool   
count  6.990280e+06  6.990280e+06  6.990280e+06  6.990280e+06  \
mean   3.748584e+00  1.184609e+00  3.265596e-01  4.986175e-01   
min    1.000000e+00 -1.000000e+00 -1.000000e+00 -1.000000e+00   
25%    3.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    4.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    5.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00   
max    5.000000e+00  1.182000e+03  7.920000e+02  4.040000e+02   
std    1.478705e+00  3.253767e+00  1.688729e+00  2.172460e+00   

                                date  
count                        6990280  
mean   2017-01-11 11:22:33.441780992  
min              2005-02-16 03:23:22  
25%    2015-01-25 04:53:50.249999872  
50%              2017-06-03 01:26:07  
75%    2019-05-23 00:02:46.249999872  
max              2022-01-19 19:48:45  
std                              NaN

In [49]:
df_review_dividido.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6990280 entries, 0 to 6990279
Data columns (total 9 columns):
 #   Column       Dtype         
---  ------       -----         
 0   review_id    object        
 1   user_id      object        
 2   business_id  object        
 3   stars        int64         
 4   useful       int64         
 5   funny        int64         
 6   cool         int64         
 7   text         object        
 8   date         datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 480.0+ MB


Solo hay valores enteros en las columnas stars, useful, funny y cool. En dates son datetime.

In [50]:
df_review_dividido.isnull().sum()

review_id      0
user_id        0
business_id    0
stars          0
useful         0
funny          0
cool           0
text           0
date           0
dtype: int64

No hay valores nulos

In [51]:
df_review_dividido["stars"].value_counts()

stars
5    3231627
4    1452918
1    1069561
3     691934
2     544240
Name: count, dtype: int64

La mayoría de valores son de votos de 5 estrellas

In [53]:
df_ordenado_por_fecha = df_review_dividido.sort_values(by="date", ascending= True)
df_ordenado_por_fecha

review_id                 user_id   
4179799  IykJMMZgbNcUndwf1q6ieg  3zBJUlWtPNoZ0uN83ODbyg  \
6295220  -O5ng1XLox6uEr4uIZ8u5A  3zBJUlWtPNoZ0uN83ODbyg   
4887456  g80vzN72iU03Wh0fSpq41g  3zBJUlWtPNoZ0uN83ODbyg   
1423159  7Dcrt0Oz0hikA8obGhadrQ  XCsZ3hWa_6oP1WkWvK7pmg   
4884223  WC9q5vhQlQkLK05kEs-vYQ  XCsZ3hWa_6oP1WkWvK7pmg   
...                         ...                     ...   
4878781  11yPqFYtkACY7TWKc5MNzA  lmiiFd9KC15fs4xtEoXRvw   
2710116  L0KaPI67DcTN4wwqfz6A3w  2Mb0st9WVyccaz6sKNLHWw   
6251618  3wqrZowVYwSPEh8p5nLn1g  3TQKP7KlNRdrI2gOkG7slg   
5476035  gUkdqUCMJCJ38HVlLZxlgg  i1PMqye40QWNkJ0MYGHuzg   
4863327  6BXQO_LAmMolJ5BYieGDsA  IH0ToaZ8hJXO2pVieN7dpQ   

                    business_id  stars  useful  funny  cool   
4179799  2bXm0SynOfxDzfrdrCyXqg      4       0      0     0  \
6295220  3g6XqkBikTTbZmTukbeGnw      4       1      0     0   
4887456  PP3BBaVxZLcJU54uP_wL6Q      5       0      0     0   
1423159  U3grYFIeu6RgAAQgdriHww      5       0      0     0   
4884223  Aes-0Q_guDeYewMapFs_vg      2       0      0     0   
...                         ...    ...     ...    ...   ...   
4878781  XDMno4l95AXgYOd0yDtHZA      5       0      0     0   
2710116  M88FFZZ2o_7QKpCFA_8RtA      5       0      0     0   
6251618  jVg-KTXEFIeAq47DTp4Hrw      5       0      0     0   
5476035  J0joPXxmN-_9Lzafspqdbw      5       0      0     0   
4863327  VItkA7pL82rCZdxHH8vBGA      5       0      0     0   

                                                      text                date  
4179799  It's not chicago style pizza but it is still a... 2005-02-16 03:23:22  
6295220  Great service.  The people really want to help... 2005-02-16 03:29:39  
4887456  These guys really are the king of cheese steak... 2005-02-16 04:06:26  
1423159  Words cannot describe how much I miss this pla... 2005-03-01 16:57:17  
4884223  Food is decent but is a complete tourist trap.... 2005-03-01 16:59:37  
...                                                    ...                 ...  
4878781  Such yummy food and great customer service! I ... 2022-01-19 19:48:13  
2710116  I LOVE Subaru of Cherry Hill.  I have bought n... 2022-01-19 19:48:16  
6251618  Jerry did an excellent job! My car is running ... 2022-01-19 19:48:19  
5476035  Love love love this place! Great food and alwa... 2022-01-19 19:48:25  
4863327  This was amongst my favorite Yelp Elite events... 2022-01-19 19:48:45  

[6990280 rows x 9 columns]

Aparentemente hay valores que datan del 2005

In [56]:
df_review_dividido["text"].duplicated().sum()

16153

Hay 16153 reviews en la columna text que estan repetidas

# EDA básico: user dataset 

In [57]:
df_user.head()

user_id    name  review_count        yelping_since  useful   
0  qVc8ODYU5SZjKXVBgXdI7w  Walker           585  2007-01-25 16:47:26    7217  \
1  j14WgRoU_-2ZE1aw1dXrJg  Daniel          4333  2009-01-25 04:35:42   43091   
2  2WnXYQFK0hXEoTxPtV2zvg   Steph           665  2008-07-25 10:41:00    2086   
3  SZDeASXq7o05mMNLshsdIA    Gwen           224  2005-11-29 04:38:33     512   
4  hA5lMy-EnncsH4JoR-hFGQ   Karen            79  2007-01-05 19:40:59      29   

   funny   cool                                              elite   
0   1259   5994                                               2007  \
1  13066  27281  2009,2010,2011,2012,2013,2014,2015,2016,2017,2...   
2   1010   1003                           2009,2010,2011,2012,2013   
3    330    299                                     2009,2010,2011   
4     15      7                                                      

                                             friends  fans  ...   
0  NSCy54eWehBJyZdG2iE84w, pe42u7DcCH2QmI81NX-8qA...   267  ...  \
1  ueRPE0CX75ePGMqOFVj6IQ, 52oH4DrRvzzl8wh5UXyU0A...  3138  ...   
2  LuO3Bn4f3rlhyHIaNfTlnA, j9B4XdHUhDfTKVecyWQgyA...    52  ...   
3  enx1vVPnfdNUdPho6PH_wg, 4wOcvMLtU6a9Lslggq74Vg...    28  ...   
4  PBK4q9KEEBHhFvSXCUirIw, 3FWPpM7KU1gXeOM_ZbYMbA...     1  ...   

   compliment_more  compliment_profile  compliment_cute  compliment_list   
0               65                  55               56               18  \
1              264                 184              157              251   
2               13                  10               17                3   
3                4                   1                6                2   
4                1                   0                0                0   

   compliment_note  compliment_plain  compliment_cool  compliment_funny   
0              232               844              467               467  \
1             1847              7054             3131              3131   
2               66                96              119               119   
3               12                16               26                26   
4                1                 1                0                 0   

   compliment_writer  compliment_photos  
0                239                180  
1               1521               1946  
2                 35                 18  
3                 10                  9  
4                  0                  0  

[5 rows x 22 columns]

In [58]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2105597 entries, 0 to 2105596
Data columns (total 22 columns):
 #   Column              Dtype  
---  ------              -----  
 0   user_id             object 
 1   name                object 
 2   review_count        int64  
 3   yelping_since       object 
 4   useful              int64  
 5   funny               int64  
 6   cool                int64  
 7   elite               object 
 8   friends             object 
 9   fans                int64  
 10  average_stars       float64
 11  compliment_hot      int64  
 12  compliment_more     int64  
 13  compliment_profile  int64  
 14  compliment_cute     int64  
 15  compliment_list     int64  
 16  compliment_note     int64  
 17  compliment_plain    int64  
 18  compliment_cool     int64  
 19  compliment_funny    int64  
 20  compliment_writer   int64  
 21  compliment_photos   int64  
dtypes: float64(1), int64(16), object(5)
memory usage: 353.4+ MB


In [61]:
df_user.isnull().sum()

user_id               0
name                  0
review_count          0
yelping_since         0
useful                0
funny                 0
cool                  0
elite                 0
friends               0
fans                  0
average_stars         0
compliment_hot        0
compliment_more       0
compliment_profile    0
compliment_cute       0
compliment_list       0
compliment_note       0
compliment_plain      0
compliment_cool       0
compliment_funny      0
compliment_writer     0
compliment_photos     0
dtype: int64

No hay valores nulos

In [63]:
df_user.duplicated().sum()

117700

117700 rows duplicadas

# EDA básico: df_tip dataset

In [1]:
df_tip.head()

NameError: name 'df_tip' is not defined

In [65]:
df_tip.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 908915 entries, 0 to 908914
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   user_id           908915 non-null  object        
 1   business_id       908915 non-null  object        
 2   text              908915 non-null  object        
 3   date              908915 non-null  datetime64[ns]
 4   compliment_count  908915 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 34.7+ MB


No tiene valores nulos

In [67]:
df_tip.duplicated().sum()

67

In [68]:
df_tip["text"].value_counts()

text
Great food                                                                                                                                                                                                                                                                                           598
Now open.                                                                                                                                                                                                                                                                                            533
Love this place!                                                                                                                                                                                                                                                                                     477
Great food!                                                                                             

Vistazo general a las reviews de la columna text. Hay al menos 851033 valores unicos.

# EDA básico: df_checkin dataset

In [69]:
df_checkin.head()

business_id                                               date
0  ---kPU91CF4Lq2-WlRu9Lw  2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020...
1  --0iUa4sNDFiZFrAdIWhZQ  2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011...
2  --30_8IhuyMHbSOcNWd6DQ           2013-06-14 23:29:17, 2014-08-13 23:20:22
3  --7PUidqRWpRSpXebiyxTg  2011-02-15 17:12:00, 2011-07-28 02:46:10, 2012...
4  --7jw19RH9JKXgFohspgQw  2014-04-21 20:42:11, 2014-04-28 21:04:46, 2014...

In [70]:
df_checkin["date"].iloc[0]

'2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020-07-24 22:42:27, 2020-10-24 21:36:13, 2020-12-09 21:23:33, 2021-01-20 17:34:57, 2021-04-30 21:02:03, 2021-05-25 21:16:54, 2021-08-06 21:08:08, 2021-10-02 15:15:42, 2021-11-11 16:23:50'

A primera vista este dataset no parece tener información muy valiosa. Contiene el id del negocio y una serie de fechas en su columna date.

In [71]:
df_checkin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131930 entries, 0 to 131929
Data columns (total 2 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   business_id  131930 non-null  object
 1   date         131930 non-null  object
dtypes: object(2)
memory usage: 2.0+ MB
